### ISSUE TITLE
Incorrect Assembly Code using >> operator and & operator

### ISSUE THREAD

**Issue by giovannipozzobon** [hth313/Calypsi-tool-chains#43]
In a MEGA65 system when writing a procedure such as the following the generated assembly code does not seem correct.

```
void setcharsetaddr(unsigned long address)
{
POKE(VIC_BASE + 0x68, (address & 0x0000FFUL));
POKE(VIC_BASE + 0x69, (address & 0xFF00UL) >> 8);
POKE(VIC_BASE + 0x6A, (address & 0xFF0000UL) >> 16);
}
```

 ```
   \ 0000          setcharsetaddr:
    \ 0000 a5..                 lda     zp:_Zp
    \ 0002 8d68d0               sta     0xd068
    \ 0005 a900                 lda     #0
    \ 0007 8d69d0               sta     0xd069
    \ 000a 8d6ad0               sta     0xd06a
    \ 000d 60                   rts


```

By modifying the code in this way instead the generated code works correctly.

```
void setcharsetaddr( unsigned long addr)
{
POKE(VIC_BASE + 0x68, (addr & 0x0000FFUL));
POKE(VIC_BASE + 0x69, ((addr >> 8) & 0x0000FFUL));
POKE(VIC_BASE + 0x6A, ((addr >> 16) & 0x0000FFUL));
}
```

also using this it should be work fine. 

void setcharsetaddr( unsigned long addr)
{
POKE(VIC_BASE + 0x68, addr); 
POKE(VIC_BASE + 0x69, (addr >> 8)); 
POKE(VIC_BASE + 0x6A, (addr >> 16));
}

Thanks.

**Comment by ki-bo** [comment]
You may want to add how POKE is defined in that example

**Comment by ki-bo** [comment]
Maybe adding the example below. One would expect that all four assignments would be compiled to just a single `lda`/`sta` pair. Also note the superfluous `sta zp:_Zp+4` in the generated code of the first assignment. The `andq` operation at the end is also not needed, as you can directly pick the MSB via `lda`.

```
void function(uint32_t a) {
  *(uint8_t *)0xd020 = a & 0x000000FFUL;
  *(uint8_t *)0xd020 = (a & 0x0000FF00UL) >> 8;
  *(uint8_t *)0xd020 = (a & 0x00FF0000UL) >> 16;
  *(uint8_t *)0xd020 = (a & 0xFF000000UL) >> 24;
}
```

```
0005                void function(uint32_t a) {
    \ 0000                      .section code,text
    \ 0000                      .public function
    \ 0000          function:
0006                  *(uint8_t *)0xd020 = a & 0x000000FFUL;
    \ 0000 a5..                 lda     zp:_Zp
    \ 0002 85..                 sta     zp:_Zp+4
    \ 0004 8d20d0               sta     0xd020
0007                  *(uint8_t *)0xd020 = (a & 0x0000FF00UL) >> 8;
    \ 0007 a900                 lda     #0
    \ 0009 8d20d0               sta     0xd020
0008                  *(uint8_t *)0xd020 = (a & 0x00FF0000UL) >> 16;
    \ 000c 8d20d0               sta     0xd020
0009                  *(uint8_t *)0xd020 = (a & 0xFF000000UL) >> 24;
    \ 000f aa                   tax
    \ 0010 a8                   tay
    \ 0011 a3ff                 ldz     #255
    \ 0013 424225..             andq    zp:_Zp
    \ 0017 424285..             stq     zp:_Zp
    \ 001b a5..                 lda     zp:_Zp+3
    \ 001d 8d20d0               sta     0xd020
0010                }
    \ 0020 60                   rts
```

**Comment by hth313** [comment]
I have fixed this for the next release.

The problem is the particular pattern `*(uint8_t *)0xd020 = (along & 0x0000FF00UL) >> 8;` (and doing the same with bytes above, expressed in this way.
As mentioned above if you rearrange the operator order it works.
I have added some additional code generator pattern to improve code generation, though there are still redundant writes in some situations that I may look into.
